pkg/gatherers/clusterconfig/node_logs.go

```diff
@@ -0,0 +1,130 @@
+
+// GatherNodeLogs fetches the node logs from journal unit
+//
+// Response see https://docs.openshift.com/container-platform/4.8/rest_api/node_apis/node-core-v1.html#apiv1nodesnameproxypath
+//
+// * Location in archive: config/nodes/logs/
+// * See: docs/insights-archive-sample/config/nodes/logs
+// * Id in config: node_logs
+func (g *Gatherer) GatherNodeLogs(ctx context.Context) ([]record.Record, []error) {
+	clientSet, err := kubernetes.NewForConfig(g.gatherProtoKubeConfig)
+	if err != nil {
+		return nil, []error{err}
+	}
+	return gatherNodeLogs(ctx, clientSet.CoreV1())
+}
+
+func gatherNodeLogs(ctx context.Context, client corev1client.CoreV1Interface) ([]record.Record, []error) {
+	nodes, err := client.Nodes().List(ctx, metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/master"})
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	var errs []error
+	records := make([]record.Record, 0)
+
+	bufferSize := int64(recorder.MaxArchiveSize * logCompressionRatio / len(nodes.Items) / 2)
+	klog.V(2).Infof("Maximum buffer size: %v bytes", bufferSize)
+	buf := bytes.NewBuffer(make([]byte, 0, bufferSize))
+
+	restClient := client.RESTClient()
+
+	for i := range nodes.Items {
+		uri := nodeLogResourceURI(restClient, nodes.Items[i].Name)
+		req := requestNodeLog(restClient, uri, logNodeMaxTailLines, logNodeUnit)
+		logString, err := nodeLogString(req, buf)
+		if err != nil {
+			klog.V(2).Infof("Error: %q", err)
+			errs = append(errs, err)
+		}
+
+		records = append(records, record.Record{
+			Name: fmt.Sprintf("config/node/logs/%s.log", nodes.Items[i].Name),
+			Item: marshal.Raw{Str: logString},
+		})
+	}
+
+	return records, errs
+}
+
+// nodeLogResourceURI creates the resource path URI to be fetched
+func nodeLogResourceURI(client rest.Interface, name string) string {
+	uri := client.Get().
+		Name(name).
+		Resource("nodes").SubResource("proxy", "logs").
+		Suffix("journal").URL().Path
+
+	if strings.HasSuffix("journal", "/") {
+		uri += "/"
+	}
+	return uri
+}
+
+// requestNodeLog creates the request to the API to retrieve the resource stream
+func requestNodeLog(client rest.Interface, uri string, tail int, unit string) *rest.Request {
+	return client.Get().RequestURI(uri).
+		SetHeader("Accept", "text/plain, */*").
+		SetHeader("Accept-Encoding", "gzip").
+		Param("tail", strconv.Itoa(tail)).
+		Param("unit", unit)
+}
+
+// nodeLogString retrieve the data from the stream, decompress it (if necessary) and return the string
+func nodeLogString(req *rest.Request, buf *bytes.Buffer) (string, error) {
+	buf.Reset()
+
+	in, err := req.Stream(context.TODO())
+	if err != nil {
+		return "", err
+	}
+	defer in.Close()
+
+	gz, err := gzip.NewReader(in)
+	if err != nil {
+		// not gzipped stream
+		_, err = io.Copy(buf, in)
+		if err != nil {
+			return "", err
+		}
+		return buf.String(), nil
+	}
+
+	for {
+		// gzipped stream
+		_, err = io.CopyN(buf, gz, 1024)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return "", err
+		}
+	}
```

Comment:
In this case, I'm handling two different possibilities. The first one that you see, using the `io.Copy` handles the non-gzipped request, so if for some reason, even asking for the gzip content, it fails to uncompress it, 'cause it is not compressed, so we can simply use the `io.Copy`. Otherwise, we have it as gzipped content, `gosec` says that there is a possibility of DDoS attacks. After some research, I found out that is recommended to use the `io.CopyN` to solve it.

Reply:
Well OK, but we don't expect a malicious content from the API server right? :) I would be in favor of ignoring the `gosec` in this case (although it's a detail again) or we can consume the stream in some different way.....

Reply:
I have no strong opinion about `io.Copy` or `io.CopyN` usage. Both works for me. Using just one could simplify the code a little bit, I guess, but I will have to split it into a separated function, I guess... 'cause we still need to handle the possibility to have the non-gzipped response, even always asking for the gzipped. Let me see what can be done.

Reply:
BTW, I added one small check to `io.EOF`, 'cause I noticed that if EOF we don't want to handle it as an error...